### PR TITLE
Add SAP System de/registration logic to Process Manager

### PR DESCRIPTION
**Pull request by jamie-suse**
# Description

Adds behaviour for processing `DatabaseInstanceRegistered`, `ApplicationInstanceRegistered`, `SapSystemRolledUp`, `DatabaseInstanceDeregistered` & `ApplicationInstanceDeregistered` events.

Adds `application_instances` & `database_instances` lists to the Process Manager state.

## How was this tested?

Added unit tests for each scenario.

### CODE REVIEW

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:271]: This logic looks wrong. You are replacing completely the database and application instances, but be aware that one host can have applications/databases from different sap systems.
So you would need to get other instances from other sap systems and concat with the ones coming in the rollup

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:185]: I got a bit lost with the order of the applies.
We should still continue having the same order like: host - cluster - sap system function. Or grouped some other way, but I found that now they are just completely unsorted

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:36]: Let's try to put a correct order. This tests are last to add, so put them at the botton of this `describe` please

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:80]: Could be use factories to create this elements?

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:150]: I miss a bit the fact the we don't test that the new instance is _added_ to the list. We always start with an empty list, which could somehow hide issues

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Hey @jamie-suse ,
Some initial comments. I think that we have some holes in our logic.
PD: I didn't finish reviewing the tests code. I will once the prod code is ready. Anyway, I would like to have a more consistent order of code additions. I wouldn't put new code in top of already existing code if there is not a logical reason.

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:67]: The new doc looks great.
So many `starts a Process Manager` look redundant.
What about:
```
- HostRegistered for a newly registered host/
- HostAddedToCluster when a Host gets added to a cluster, as this event may arrive prior to ...
- DatabaseInstanceRegistered when an instance gets...
```

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:159]: We still have the scenario where this `cluster_id` could be nil. We need to take care of that.

File: test/support/factory.ex
Comment (arbulu89) [inline test/support/factory.ex:359]: `Health.passing()`

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:91]: Let's be more generous here and create a list of instances. At the end, we do operations with them

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:260]: remove `after`

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:266]: Maybe add at least one instance to database_instance or application_instance, to see that it adds the values properly

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Hey @jamie-suse ,
in general terms it look great.
Only missing to handle properly some other scenario when the cluster is not there but app/database are

**Review by rtorrero — APPROVED** [review]
LGTM

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:126]: With the next function this one is not needed anymore, right?

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:157]: What about:
```
List.flatten([
    Enum.map(database_instances, fn %Instance{
    ...
    end),
    Enum.map(application_instances, fn %Instance{
    ...
    end),
    if cluster_id do
        %DeregisterClusterHost{...
    end,
    %DeregisterHost{...
])
```
This way we respect the logical order of commands as well
The `if` condition smells a bit and maybe there is a better way of doing this with pattern matching.

File: test/trento/application/process_managers/deregistration_process_manager_test.exs
Comment (arbulu89) [inline test/trento/application/process_managers/deregistration_process_manager_test.exs:92]: This looks an antipattern:
Maybe:
```
[%{host_id: db_host_id_1}, %{host_id: db_host_id_2}] = database_instances = build_list(2, :sap_system_instance)
```
And so on. In this case (and almost always when you use factories), giving pre-created unique id-s to the created objects doesn't make sense. Let the factory do that work for you. You can capture later on these values.

**Review by arbulu89 — CHANGES_REQUESTED** [review]
Some final improvements from my side

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (jamie-suse) [reply lib/trento/application/process_managers/deregistration_process_manager.ex:126]: 🧠 Ah yes, I missed that

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [inline lib/trento/application/process_managers/deregistration_process_manager.ex:128]: So now we have 40 lines function duplicated to avoid that `if` XD
I even preferred the `if` option honestly

**Review by arbulu89 — APPROVED** [review]
Green from my light.
I have to confess that i might miss something, as following the history after so many reviews with overriding commits is hard.  At the end, I don't know what was changed after every commit.
I would recommend to wait until receiving other approve at least

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (arbulu89) [reply lib/trento/application/process_managers/deregistration_process_manager.ex:128]: Or maybe some auxiliar function to flatten the app/db instances so it can be reused in the current code. At the end, that part of the code is duplicated

**Review by fabriziosestito — CHANGES_REQUESTED** [review]

File: lib/trento/application/process_managers/deregistration_process_manager.ex
Comment (fabriziosestito) [reply lib/trento/application/process_managers/deregistration_process_manager.ex:25]: @CDimonaco could you confirm?

Comment (CDimonaco) [reply lib/trento/application/process_managers/deregistration_process_manager.ex:25]: Yes this should be empty, I confirm

**Review by CDimonaco — APPROVED** [review]

**Review by arbulu89 — APPROVED** [review]

**Pull request merged**
